src/torsion_profile.py

```diff
@@ -0,0 +1,191 @@
+""" @package forcebalance.opt_geo_target Optimized Geometry fitting module.
+
+@author Yudong Qiu, Lee-Ping Wang
+@date 03/2019
+"""
+from __future__ import division
+import os
+import shutil
+import numpy as np
+import re
+import subprocess
+import itertools
+import json
+from copy import deepcopy
+from collections import OrderedDict, defaultdict
+from forcebalance.nifty import col, eqcgmx, flat, floatornan, fqcgmx, invert_svd, kb, printcool, printcool_dictionary, bohr2ang, warn_press_key
+from forcebalance.target import Target
+from forcebalance.molecule import Molecule
+from forcebalance.finite_difference import fdwrap, f1d2p, f12d3p, in_fd
+from forcebalance.output import getLogger
+from forcebalance.optimizer import Counter
+from forcebalance.opt_geo_target import periodic_diff
+logger = getLogger(__name__)
+
+class TorsionProfileTarget(Target):
+    """ Subclass of Target for fitting MM optimized geometries to QM optimized geometries. """
+    def __init__(self, options, tgt_opts, forcefield):
+        super(TorsionProfileTarget, self).__init__(options, tgt_opts, forcefield)
+
+        if not os.path.exists(os.path.join(self.tgtdir, 'metadata.json')):
+            raise RuntimeError('TorsionProfileTarget needs torsion drive metadata.json file')
+        with open(os.path.join(self.tgtdir, 'metadata.json')) as f:
+            self.metadata = json.load(f)
+            self.ndim = len(self.metadata['torsion_grid_ids'][0])
+            self.freeze_atoms = sorted(list(set(itertools.chain(*self.metadata['dihedrals']))))
+        
+        ## Read in the coordinate files and get topology information from PDB
+        if hasattr(self, 'pdb') and self.pdb is not None:
+            self.mol = Molecule(os.path.join(self.root,self.tgtdir,self.coords),
+                                top=(os.path.join(self.root,self.tgtdir,self.pdb)))
+        else:
+            self.mol = Molecule(os.path.join(self.root,self.tgtdir,self.coords))
+        ## Number of snapshots.
+        self.ns = len(self.mol)
+        ## Option for how much data to write to disk.
+        self.set_option(tgt_opts,'writelevel','writelevel')
+        ## Harmonic restraint for non-torsion atoms in kcal/mol.
+        self.set_option(tgt_opts,'restrain_k','restrain_k')
+        ## Attenuate the weights as a function of energy
+        self.set_option(tgt_opts,'attenuate','attenuate')
+        ## Energy denominator for objective function
+        self.set_option(tgt_opts,'energy_denom','energy_denom')
+        ## Set upper cutoff energy
+        self.set_option(tgt_opts,'energy_upper','energy_upper')
+        ## Read in the reference data.
+        self.read_reference_data()
+        ## Build keyword dictionaries to pass to engine.
+        engine_args = OrderedDict(list(self.OptionDict.items()) + list(options.items()))
+        del engine_args['name']
```

Comment:
Small question: `del engine_args['name']` will raise `KeyError` if `engine_args` does not have `'name'` key. Is this what we want to check here? If not I think we can do
```
engine_args.pop('name', None)
```
So it won't raise Error.

Reply:
Right, this problem is because the engine and target options both have the "name" key, and the target name should override the engine name.  A better design would have been to rename the Engine "name" key to something like "engine_name".  We should do that at some point.